### PR TITLE
#55 Update multiple fields in indicate_errors

### LINKED ISSUE #55: Multiple variables: only one field being updated at a time in goal based indicate_errors
Hey @jwallwork23 and @stephankramer. As `GoalOrientedMeshSeq.indicate_errors()` is currently written, we are updating fields used in computing error indicators by first looping through fields and then looping through solutions at exported timesteps ([L194-L201](https://github.com/pyroteus/pyroteus/blob/5a261d8488e36931e6cabb12a8356795cf9e96d4/pyroteus/go_mesh_seq.py#L194-L201)). This means that if we have more than one field in our mesh sequence, the updated fields won't represent the same timestep since only one field is being updated at the current timestep in the loop. This is an issue, for example, in the Gray-Scott example that we have.

In this multiple variable case, is it enough to just update all fields at the current timestep, something like this (just adding another loop)
```
for f in self.fields:
    # Loop over each timestep
    for j in range(len(sols[f]["forward"][i])):
        # Update fields
        for ff in self.fields:
          transfer(sols[ff][FWD][i][j], u[ff])
          transfer(sols[ff][FWD_OLD][i][j], u_[ff])
          transfer(sols[ff][ADJ][i][j], u_star[ff])
          transfer(sols[ff][ADJ_NEXT][i][j], u_star_next[ff])
```
or do we also need to modify the other calculations further down the line?

### PR DESCRIPTION
**Pull request by ddundo**
Closes #55.

@acse-ej321 could you please take a look as well? I'm not confident how clear the long comment block is, so I'd appreciate a fresh set of eyes to take a look, since Joe and I are already familiar with the issue :)

### CODE REVIEW

**Comment by ddundo** [comment]
@jwallwork23 I didn't want to do this on your [55_multiple_var_ee](https://github.com/mesh-adaptation/goalie/tree/55_multiple_var_ee) branch. I left it to you to delete :)

File: goalie/go_mesh_seq.py [hunk 1]
```diff
@@ -185,17 +185,24 @@ def indicate_errors(
...
+            for j in range(self.time_partition.num_exports_per_subinterval[i] - 1):
+                # Loop over each strongly coupled field
+                for f_idx, f in enumerate(self.fields):
+                    # Transfer solutions associated with the current field f
                     transfer(self.solutions[f][FWD][i][j], u[f])
                     transfer(self.solutions[f][FWD_OLD][i][j], u_[f])
                     transfer(self.solutions[f][ADJ][i][j], u_star[f])
                     transfer(self.solutions[f][ADJ_NEXT][i][j], u_star_next[f])
+                    # In case of multiple solution fields, the field that is solved for
+                    # first used the values of other fields from the previous timestep.
```
Comment (jwallwork23) [inline goalie/go_mesh_seq.py:198]: Do you mean "uses" rather than "used"?

File: goalie/go_mesh_seq.py [hunk 2]
```diff
@@ -185,17 +185,24 @@ def indicate_errors(
...
+                    # Therefore, transfer the lagged solution of fields f_next that are
+                    # solved for after the current field f. This assumes that the order
+                    # of fields being solved for in get_solver is the same as their
+                    # order in self.fields
```
Comment (jwallwork23) [inline goalie/go_mesh_seq.py:202]: Thanks for mentioning this assumption explicitly.

File: goalie/go_mesh_seq.py [hunk 3]
```diff
@@ -185,17 +185,24 @@ def indicate_errors(
...
+                    if f_idx < len(self.fields) - 1:
+                        for f_next in self.fields[f_idx + 1 :]:
+                            transfer(self.solutions[f_next][FWD_OLD][i][j], u[f_next])
```
Comment (jwallwork23) [inline goalie/go_mesh_seq.py:205]: Is there not repeated effort here? e.g., doesn't the final `FWD_OLD` solution get updated with the same values several times? Could we not just update the `FWD_OLD` solutions for all fields *before* the loop on L191?

**Review by jwallwork23 — CHANGES_REQUESTED** [review]

File: goalie/go_mesh_seq.py (hunk 3 above)
Comment (ddundo) [reply goalie/go_mesh_seq.py]: Ah good point, thanks! Fixed in 1e88b2b0f303b0904fc286c35663774575a55ebc

File: goalie/go_mesh_seq.py (hunk 1 above)
Comment (ddundo) [reply goalie/go_mesh_seq.py]: I put "used" to try to emphasise that I am referring to get_solver :) I changed to "uses" in 1e88b2b0f303b0904fc286c35663774575a55ebc but I'm still not sure how clear this comment block is in general.

**Review by jwallwork23 — APPROVED** [review]
Thanks. Why did you cancel the CI? Please trigger it so that we can check the tests pass.

**Comment by ddundo** [comment]
> Thanks. Why did you cancel the CI? Please trigger it so that we can check the tests pass.

Because tests fail anyway since the docker image doesn't have the `create_section` fix :)

**Comment by jwallwork23** [comment]
I've rebuild the Docker image, will trigger tests now.

**Comment by jwallwork23** [comment]
Hm, there seems to be a problem with `assemble_mixed_mass_matrix` after updating. Will look into it.

**Comment by ddundo** [comment]
> Hm, there seems to be a problem with `assemble_mixed_mass_matrix` after updating. Will look into it.

Thanks Joe! Everything passed for me yesterday after updating firedrake on Monday, but I updated it again now and get the same tests failing. Sorry about that.

**Comment by acse-ej321** [comment]
@jwallwork23 @ddundo - I am getting an error with `assemble_mixed_mass_matrix` passing the `load` instructions to pyop2 without an MPI `comm` argument. It looks a test for this was added  [in pyop2 within the last day](https://github.com/OP2/PyOP2/commit/7bef38fa5c255d74379fcfe50302da5f654ce2a4) might explain why the tests were not failing before and are now. 
I think I have it working with a local hack of [Firedrake,](https://github.com/firedrakeproject/firedrake/blob/af2c9cc92ff295fdc2916978220c6a148cb26741/firedrake/supermeshing.py#L434-L438) adding a `comm=COMM_SELF` argument to the `load` call from `assemble_mixed_mass_matrix`. I am not sure this is the correct solution though.

**Comment by ddundo** [comment]
Thanks for looking into this anyway @acse-ej321 :)

**Comment by acse-ej321** [comment]
@jwallwork23 - It appears the fix for this has been merged into firedrake `main` now.

**Comment by jwallwork23** [comment]
> I've rebuild the Docker image, will trigger tests now.

And again. Now fixed. Ready for merge if you're happy @ddundo

**Comment by ddundo** [comment]
Thanks both!

**Pull request merged**